### PR TITLE
diff: add support for absolute tolerance of floats

### PR DESCRIPTION
**Pull request by adrien-berchet**
Hi
I encountered an issue using this nice package: when I want to compare floats, I sometimes need absolute tolerance because of very small values. For example, I don't really care about 100% difference between `1e-16` and `2e-16`, while I do care of a 1% difference between `1` and `1.01`. This case could be solved using a tolerance of `0.01` and an absolute tolerance of `1e-15`.
This PR adds this feature with a few test cases that show more examples.

### CODE REVIEW

File: dictdiffer/__init__.py [hunk 1]
```diff
@@ -35,7 +35,7 @@
 
 
 def diff(first, second, node=None, ignore=None, path_limit=None, expand=False,
-         tolerance=EPSILON, dot_notation=True):
+         tolerance=EPSILON, absolute_tolerance=None, dot_notation=True):
```
Comment (joesolly) [inline dictdiffer/__init__.py:38]: looking at the code below with the `rel_diff and abs_diff`, it looks like you could get into a situation where you always have a diff if both tolerances are set to `None`. What if we reuse tolerance and add a parameter called `tolerance_type='relative'` with the other option being `'absolute'` (or an ENUM). Then we can have a single comparison unless you can think of a use case for having both comparisons.

File: dictdiffer/utils.py [hunk 2]
```diff
@@ -270,6 +270,17 @@ def are_different(first, second, tolerance):
         return not (first_is_nan and second_is_nan)
     elif isinstance(first, num_types) and isinstance(second, num_types):
         # two numerical values are compared with tolerance
-        return abs(first-second) > tolerance * max(abs(first), abs(second))
+        if tolerance is not None:
+            rel_diff = (
+                abs(first-second) > tolerance * max(abs(first), abs(second))
+            )
+        else:
+            rel_diff = True
```
Comment (joesolly) [inline dictdiffer/utils.py:278]: shouldn't this be `False` and the return statement below be an `or`?

File: dictdiffer/utils.py [hunk 3]
```diff
@@ -270,6 +270,17 @@ def are_different(first, second, tolerance):
...
+        # two numerical values are compared with absolute_tolerance
+        if absolute_tolerance is not None:
+            abs_diff = (abs(first-second) > absolute_tolerance)
+        else:
+            abs_diff = True
+        return rel_diff and abs_diff
```
Comment (joesolly) [inline dictdiffer/utils.py:284]: we could add a third case here where there is no tolerance (or absolute_tolerance=0 which would require the above comparison to become >=) so that you can allow only exact matches.

**Review by joesolly — COMMENTED** [review]
This is a pretty awesome change 👍

**Comment by momirza** [comment]
😆 

Didn't realise this existed so basically implemented the similar changes... 

https://github.com/inveniosoftware/dictdiffer/pull/154

File: dictdiffer/__init__.py (hunk 1 above)
Comment (adrien-berchet) [reply dictdiffer/__init__.py:38]: If both tolerances are `None`, then only the strictly equal numbers will be considered as equal thanks to line 262: `if first == second`.
I prefer having both comparison since I want to use the absolute threshold to ignore very small numbers and compare only the relevant ones.
For the interface I was inspired by the `numpy.isclose()` function which I find easy to understand and to use.

File: dictdiffer/utils.py (hunk 3 above)
Comment (joesolly) [reply dictdiffer/utils.py:284]: Ignore this as I missed line 262.

File: dictdiffer/utils.py (hunk 2 above)
Comment (joesolly) [reply dictdiffer/utils.py:278]: I think this is more preferential than required, but give it a thought. It seems like either way gives the same result.

Comment (adrien-berchet) [reply dictdiffer/utils.py:278]: It depends on what you want :-)
I tried to add all the relevant tests to illustrate the cases that are covered by this feature. You can check them in `test_dictdiffer.py` (mainly between line 117 and line 138 for this specific topic). As you can see on line 120, if the absolute tolerance is satisfied, then they are considered as equal, even if the relative tolerance is not satisfied.This is exactly what I wanted to achieve but it might not be what you want?

File: dictdiffer/utils.py (hunk 3 above)
Comment (adrien-berchet) [reply dictdiffer/utils.py:284]: If I understand your point correctly, this case is already covered (thanks to `utils.py:262`), as shown by the test line 150.

**Comment by adrien-berchet** [comment]
For the failing Checks, I think it is not related to this PR since it fails during the dependencies installation step. This should be fixed in another PR and then I will rebase to make the Checks pass.

File: dictdiffer/utils.py (hunk 2 above)
Comment (joesolly) [reply dictdiffer/utils.py:278]: I take back my suggestion. We are talking about the same outcome, I was reversing my expression, but my suggestion wouldn't work when both tolerances are set to `None`. Sorry about that!

**Comment by adrien-berchet** [comment]
> laughing
> 
> Didn't realise this existed so basically implemented similar changes...
> 
> #154

Hehe, this means that this feature is a must have ;-)
We did not use the exact same equation to compute the difference, we could discuss which one should be prefered.

File: dictdiffer/utils.py (hunk 2 above)
Comment (adrien-berchet) [reply dictdiffer/utils.py:278]: np :)

File: dictdiffer/utils.py (hunk 3 above)
Comment (adrien-berchet) [inline dictdiffer/utils.py:284]: I just realized that what we want already exists in `math`, so we could just write this:

```suggestion
    return not math.isclose(
            first,
            second,
            rel_tol=tolerance or 0,
            abs_tol=absolute_tolerance or 0,
        )
```
It is easier to understand, already documented and probably more efficient.

Comment (momirza) [reply dictdiffer/utils.py]: This is much better! I'll close my PR in favour of this.

**Comment by adrien-berchet** [comment]
Hello @jirikuncar 
Now that #158 is merged, could we try to merge this one and build a new release please?

File: dictdiffer/utils.py [hunk 4]
```diff
@@ -270,6 +271,11 @@ def are_different(first, second, tolerance):
         return not (first_is_nan and second_is_nan)
     elif isinstance(first, num_types) and isinstance(second, num_types):
         # two numerical values are compared with tolerance
-        return abs(first-second) > tolerance * max(abs(first), abs(second))
+        return not math.isclose(
```
Comment (jirikuncar) [inline dictdiffer/utils.py:274]: Can you please remove PY 2.7 from `setup.py` and make sure we support only PY 3.5+?

Comment (adrien-berchet) [reply dictdiffer/utils.py:274]: Ah yes, sure. Sorry I didn't see that `math.isclose` was only introduced in py35.
It should be good now. I had to set a constraint on the selenium version to make it work locally but I am not sure it is mandatory.

Comment (mikaelho) [reply dictdiffer/utils.py:274]: Is the `_compat.py` then obsolete as well?

Comment (adrien-berchet) [reply dictdiffer/utils.py:274]: I cleaned a few things in `_compat.py` and tested manually with py35. It needs a few tweaks to work with py35 so I did not include it in the CI (one must use `pytest-invenio<1.4.0` and change the `addopts` in `pytest.ini`).

**Comment by adrien-berchet** [comment]
Can you trigger the CI please? So I can see if it can pas now.

File: dictdiffer/_compat.py [hunk 5]
```diff
@@ -5,27 +5,9 @@
...
-    from itertools import izip_longest as _zip_longest
-else:
-    from collections import (Iterable, MutableMapping, MutableSequence,
-                             MutableSet)
-    from itertools import zip_longest as _zip_longest
-
-izip_longest = _zip_longest
+string_types = str,
+text_type = str
+num_types = int, float
```
Comment (jirikuncar) [inline dictdiffer/_compat.py:13]: We should probably completely remove this file. It would be nice to do this change in a separate PR which would remove the support for PY<3.5.

File: dictdiffer/utils.py [hunk 6]
```diff
@@ -252,7 +253,7 @@ def dot_lookup(source, lookup, parent=False):
     return value
 
 
-def are_different(first, second, tolerance):
+def are_different(first, second, tolerance, absolute_tolerance):
```
Comment (jirikuncar) [inline dictdiffer/utils.py:256]: Can we add a default value for the `absolute_tolerance` parameter to avoid breaking change?

Comment (adrien-berchet) [reply dictdiffer/utils.py]: Sure, done.

File: dictdiffer/_compat.py (hunk 5 above)
Comment (adrien-berchet) [reply dictdiffer/_compat.py:13]: Yeah, as you prefer.
Do you want me to revert this file?

Comment (jirikuncar) [reply dictdiffer/_compat.py:13]: Revert here or remove in separate PR as you prefer.

Comment (adrien-berchet) [reply dictdiffer/_compat.py:13]: Ok, I am going to do another PR to drop support of Py<3.5 then I will rebase this one on it.

Comment (adrien-berchet) [reply dictdiffer/_compat.py:13]: See #160

File: dictdiffer/utils.py (hunk 4 above)
Comment (jirikuncar) [reply dictdiffer/utils.py:274]: I would say we should remove the dependency in `pytest-invenio` and list the packages manually.

Comment (adrien-berchet) [reply dictdiffer/utils.py:274]: If we don't use any feature of `pytest-invenio` is used it would be better indeed. I am going to check that.

**Comment by adrien-berchet** [comment]
@jirikuncar it looks good to me, I let you check.

**Review by jirikuncar — APPROVED** [review]

**Comment by adrien-berchet** [comment]
Thanks!
Could you plan a new release soon please?
